Lib/test/test_type_comments.py

```diff
@@ -0,0 +1,203 @@
+import ast
+import unittest
+
+
+funcdef = """\
+def foo():
+    # type: () -> int
+    pass
+
+def bar():  # type: () -> None
+    pass
+"""
+
+asyncdef = """\
+async def foo():
+    # type: () -> int
+    return await bar()
+
+async def bar():  # type: () -> int
+    return await bar()
+"""
+
+forstmt = """\
+for a in []:  # type: int
+    pass
+"""
+
+
+withstmt = """\
+with context():  # type: int
+    pass
+"""
+
+vardecl = """\
+a = 0  # type: int
+a  # type: int
```

Comment:
per your comment in the test below regarding this being a curiosity...  do we want such lenient behavior?  if we leave whether or not type comments in "odd" places are stored as an implementation details, something will eventually wind up depending on our specific implementation of the undefined behavior.

Reply:
I don't have a strong opinion but I would rather make it a `SyntaxError`.